# Implementation

## Imports

In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import os

In [2]:
plt.style.use('dark_background')

In [3]:
import sklearn
import keras
import tensorflow as tf

In [4]:
from keras.applications import InceptionV3

In [5]:
import cv2

In [6]:
path = os.getcwd()
path = path[:-4]
data_dir = path + "/data/lfw-deepfunneled"

In [7]:
people = pd.read_csv(path + "/data/people.csv")
people = pd.read_csv(path + "/data/people.csv")
names = [item for item in people["name"].tolist() if item in os.listdir(data_dir)]
people = people[people["name"].isin(names)]
N = 10
people = people[people.images > N]
k = len(people)
print(f"k = {k} classes")

k = 143 classes


In [8]:
inceptionv3 = InceptionV3()
model = InceptionV3(include_top=False, classes=k, pooling='avg', weights='imagenet')
model = keras.Sequential(layers=(model, keras.layers.Dense(k, activation='softmax')))

## Data Preparation

In [9]:
people = people.rename(columns={'images': 'num_images'})

In [10]:
people = people.set_index('name')

In [11]:
image_col = []
for person in people.iterrows():
    person = person[1]
    image_list = []
    for i in range(int(person["num_images"])):
        image_path = data_dir + '/' + person.name + '/' + person.name + '_' +  "{:04d}".format(i+1) + '.jpg'
        image = cv2.imread(image_path)
        image = cv2.resize(image, (299, 299))
        #image = tf.expand_dims(image, axis=0)
        image_list.append(image)
    image_col.append(image_list)

In [12]:
people["images"] = image_col

In [13]:
# num_images == len(images) for each person
assert ([len(x) for x in people["images"]] == people["num_images"].values).all()

In [14]:
dataset = [] # element: (image, person, train or test)
for person in people.iterrows():
    person = person[1]
    image_list = person["images"]
    num_images = len(image_list)
    for i in range(num_images):
        image = image_list[i]
        element = [image, person.name, "test" if i>N-2 else "train"]
        dataset.append(element)
dataset = pd.DataFrame(dataset, columns=["image", "person", "split"])

In [15]:
dataset = pd.get_dummies(dataset, columns=['person'])

In [16]:
dataset.head()

image  split  \
0  [[[0, 1, 0], [0, 1, 0], [0, 1, 0], [0, 1, 0], ...  train   
1  [[[0, 1, 1], [0, 1, 1], [0, 1, 1], [0, 1, 1], ...  train   
2  [[[0, 0, 0], [0, 0, 0], [1, 0, 0], [2, 0, 0], ...  train   
3  [[[0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0], ...  train   
4  [[[129, 163, 176], [129, 163, 176], [129, 163,...  train   

   person_Abdullah_Gul  person_Adrien_Brody  person_Alejandro_Toledo  \
0                False                False                    False   
1                False                False                    False   
2                False                False                    False   
3                False                False                    False   
4                False                False                    False   

   person_Alvaro_Uribe  person_Amelie_Mauresmo  person_Andre_Agassi  \
0                False                   False                False   
1                False                   False                False   
2                False                   False                False   
3                False                   False                False   
4                False                   False                False   

   person_Andy_Roddick  person_Angelina_Jolie  ...  person_Tom_Ridge  \
0                False                  False  ...             False   
1                False                  False  ...             False   
2                False                  False  ...             False   
3                False                  False  ...             False   
4                False                  False  ...             False   

   person_Tommy_Franks  person_Tony_Blair  person_Trent_Lott  \
0                False              False              False   
1                False              False              False   
2                False              False              False   
3                False              False              False   
4                False              False              False   

   person_Venus_Williams  person_Vicente_Fox  person_Vladimir_Putin  \
0                  False               False                  False   
1                  False               False                  False   
2                  False               False                  False   
3                  False               False                  False   
4                  False               False                  False   

   person_Wen_Jiabao  person_Winona_Ryder  person_Yoriko_Kawaguchi  
0              False                False                    False  
1              False                False                    False  
2              False                False                    False  
3              False                False                    False  
4              False                False                    False  

[5 rows x 145 columns]

## Model Training on LFW Dataset

<img src="fig2.png" alt="drawing" width="600"/><br>
1. Train model using N-1 images of each person.
2. Use remaining images as test dataset, get results.

In [17]:
X_train = dataset[dataset["split"] == "train"]["image"].tolist()
X_train = np.asarray(X_train).astype('float32')
X_test = dataset[dataset["split"] == "test"]["image"].tolist()
X_test = np.asarray(X_test).astype('float32')

In [18]:
y_train = dataset[dataset["split"] == "train"].drop(['image', 'split'], axis=1)
y_test = dataset[dataset["split"] == "test"].drop(['image', 'split'], axis=1)

In [19]:
model.compile(loss="categorical_crossentropy", optimizer=keras.optimizers.Adam(learning_rate=1e-4), metrics=['accuracy'])

In [20]:
model.fit(X_train, y_train, shuffle=True, epochs=10)

Epoch 1/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 290s 7s/step - accuracy: 0.0285 - loss: 4.9700
Epoch 2/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 293s 7s/step - accuracy: 0.7763 - loss: 3.1722
Epoch 3/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 309s 8s/step - accuracy: 0.9829 - loss: 1.6962
Epoch 4/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 317s 8s/step - accuracy: 0.9973 - loss: 0.7308
Epoch 5/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 315s 8s/step - accuracy: 1.0000 - loss: 0.2784
Epoch 6/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 313s 8s/step - accuracy: 1.0000 - loss: 0.1290
Epoch 7/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 314s 8s/step - accuracy: 1.0000 - loss: 0.0769
Epoch 8/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 310s 8s/step - accuracy: 1.0000 - loss: 0.0548
Epoch 9/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 310s 8s/step - accuracy: 0.9978 - loss: 0.0538
Epoch 10/10
41/41 ━━━━━━━━━━━━━━━━━━━━ 308s 8s/step - accuracy: 1.0000 - loss: 0.0376


In [21]:
p = model(tf.expand_dims(X_train[0], axis=0))

In [22]:
np.argmax(p)

8

In [23]:
np.argmax(y_train.iloc[0])

8

In [24]:
p = model(tf.expand_dims(X_test[0], axis=0))

In [25]:
np.argmax(p)

8

In [26]:
np.argmax(y_test.iloc[0])

8

In [27]:
model.evaluate(X_test, y_test)

91/91 ━━━━━━━━━━━━━━━━━━━━ 145s 2s/step - accuracy: 0.4937 - loss: 2.3526


[2.4133005142211914, 0.46553516387939453]

## Conclusion
For $k=143$ classes, each with at least $N=10$ image samples, and leave-one-out test split schema, yield overfitted $100\%$ train accuracy and $46\%$ test accuracy.